File: app/src/main/java/shetj/me/base/common/other/GlideImageLoader.java
package shetj.me.base.common.other;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.transition.Transition;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.reactivex.Flowable;
import io.reactivex.SingleObserver;
import io.reactivex.schedulers.Schedulers;
import me.shetj.base.tools.app.Utils;
import shetj.me.base.view.LoadingDialog;
import shetj.me.base.utils.ImageLoader;

/**
 * <b>@packageName：</b> me.shetj.base.base<br>
 * <b>@author：</b> shetj<br>
 * <b>@createTime：</b> 2018/10/29 0029<br>
 * <b>@company：</b><br>
 * <b>@email：</b> 375105540@qq.com<br>
 * <b>@describe</b> 具体用法 {@https://muyangmin.github.io/glide-docs-cn/doc/transitions.html}<br>
 *   加载图片三步走，1,with; 2,load; 3 into;
 *   取消图片也是三步走，1,with; 2,load; 3 clear;
 */
public class GlideImageLoader implements ImageLoader {
	@Override
	public void load(@NonNull ImageView simpleView,@NonNull String url) {
		GlideApp.with(simpleView.getContext())
						.load(url)
						.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
						.into(simpleView);
	}

	@Override
	public void load(@NonNull ImageView simpleView,@NonNull String url,@NonNull boolean hasProgress) {
			if (hasProgress){
				GlideApp.with(simpleView.getContext())
								.load(url)
								.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
								.into(new SimpleTarget<Drawable>() {
									@Override
									public void onStart() {
										super.onStart();
										LoadingDialog.showLoading((Activity) simpleView.getContext());
									}

									@Override
									public void onResourceReady(@NonNull Drawable resource, @Nullable Transition<? super Drawable> transition) {
										simpleView.setImageDrawable(resource);
										LoadingDialog.hideLoading();
									}
								});
			}else {
				load(simpleView,url);
			}
	}

	/**
	 * 未写
	 * @param mSimpleView 容器
	 * @param url url地址
	 */
	@Override
	public void loadProgressive(@NonNull ImageView mSimpleView, @NonNull String url) {
		GlideApp.with(mSimpleView.getContext())
						.load(url)
						.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
						.into(mSimpleView);
	}

	@Override
	public void loadGif(@NonNull ImageView simpleView,@NonNull String url,@NonNull boolean isAuto) {
		GlideApp.with(simpleView.getContext())
						.load(url)
						.diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
						.into(simpleView);
	}

	@Override
	public void prefetchImage(final Context context,@NonNull final String url, boolean isDiskCacheOrBitmapCache) {
		Flowable.just(url)
						.isEmpty()
						.observeOn(Schedulers.io())
						.subscribe(it ->
										Glide.with(context)
										.load(it)
										.submit());
	}

	@Override
	public ImageView getSimpleView(Context context, String url) {
		ImageView imageView = new ImageView(context);
		load(imageView,url);
		return   imageView;
	}

	@Override
	public void clearMemCache() {
		Glide.get(Utils.Companion.getApp().getApplicationContext()).clearMemory();
	

	}

	@Override
	public void clearCacheFiles() {
		Flowable.just(1).observeOn(Schedulers.newThread())
						.subscribe(integer -> {
							//清理磁盘缓存 需要在子线程中执行
							Glide.get(Utils.Companion.getApp().getApplicationContext()).clearDiskCache();
						});
	}

	@Override
	public void displayImage(@NotNull Context context, @NotNull String url, @NotNull ImageView view) {

	}

	@Override
	public void displayUserImage(@NotNull Context context, @NotNull String url, @NotNull ImageView view) {

	}

	@Override
	public void preLoad(@NotNull Context context, @NotNull String url, @NotNull ImageView view) {

	}
}

File: app/src/main/java/shetj/me/base/common/other/GlideImageLoader.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 SheTieJun
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


package shetj.me.base.common.other

import android.content.Context
import android.widget.ImageView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.bitmap.DownsampleStrategy
import io.reactivex.rxjava3.schedulers.Schedulers
import me.shetj.base.tools.app.Utils.Companion.app
import shetj.me.base.utils.ImageLoader

/**
 * **@packageName：** me.shetj.base.base<br></br>
 * **@author：** shetj<br></br>
 * **@createTime：** 2018/10/29 0029<br></br>
 * **@company：**<br></br>
 * **@email：** 375105540@qq.com<br></br>
 * **@describe**
 * 具体用法 {@https://muyangmin.github.io/glide-docs-cn/doc/transitions.html}<br></br>
 * 加载图片三步走，1,with; 2,load; 3 into;
 * 取消图片也是三步走，1,with; 2,load; 3 clear;
 */
class GlideImageLoader private constructor() : ImageLoader {

    companion object {

        @Volatile
        private var instance: GlideImageLoader? = null

        fun getInstance(): GlideImageLoader {
            return instance ?: synchronized(GlideImageLoader::class.java) {
                GlideImageLoader().also {
                    instance = it
                }
            }
        }
    }


    override fun load(simpleView: ImageView, url: String) {
        Glide.with(simpleView.context)
                .load(url)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .into(simpleView)
    }

    override fun load(simpleView: ImageView, url: String, hasProgress: Boolean) {
            load(simpleView, url)
    }

    /**
     * 未写
     * @param mSimpleView 容器
     * @param url url地址
     */
    override fun loadProgressive(mSimpleView: ImageView, url: String) {
        Glide.with(mSimpleView.context)
                .load(url)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .into(mSimpleView)
    }

    override fun loadGif(simpleView: ImageView, url: String, isAuto: Boolean) {
        Glide.with(simpleView.context)
                .load(url)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .downsample(DownsampleStrategy.DEFAULT)
                .into(simpleView)
    }

    override fun prefetchImage(context: Context, url: String, isDiskCacheOrBitmapCache: Boolean) {
        Glide.with(context)
                .load(url)
                .submit()
    }

    override fun getSimpleView(context: Context, url: String): ImageView {
        val imageView = ImageView(context)
        load(imageView, url)
        return imageView
    }

    override fun clearMemCache() {
        Glide.get(app.applicationContext).clearMemory()
    }

    override fun clearCacheFiles() {
        Schedulers.io().scheduleDirect {
            Glide.get(app.applicationContext).clearDiskCache()
        }
    }

    override fun displayImage(context: Context, url: String, view: ImageView) {}
    override fun displayUserImage(context: Context, url: String, view: ImageView) {}
    override fun preLoad(context: Context, url: String) {}
}